core-primitives/stf-state-handler/src/file_io.rs

```diff
@@ -18,134 +18,284 @@
 #[cfg(all(not(feature = "std"), feature = "sgx"))]
 use crate::sgx_reexport_prelude::*;
 
-use crate::error::{Error, Result};
+#[cfg(feature = "std")]
+use rust_base58::base58::{FromBase58, ToBase58};
+
+#[cfg(feature = "sgx")]
 use base58::{FromBase58, ToBase58};
+
+use crate::error::{Error, Result};
 use codec::{Decode, Encode};
-use ita_stf::{State as StfState, StateType as StfStateType, Stf};
-use itp_settings::files::{ENCRYPTED_STATE_FILE, SHARDS_PATH};
-use itp_sgx_crypto::{AesSeal, StateCrypto};
-use itp_sgx_io::{read as io_read, write as io_write, SealedIO};
-use itp_types::{ShardIdentifier, H256};
-use log::*;
-use sgx_tcrypto::rsgx_sha256_slice;
-use sgx_types::sgx_status_t;
-use std::{format, fs, io::Write, path::Path, vec::Vec};
-
-pub(crate) fn load_initialized_state(shard: &ShardIdentifier) -> Result<StfState> {
-	trace!("Loading state from shard {:?}", shard);
-	let state = if exists(&shard) {
-		load(&shard)?
-	} else {
-		trace!("Initialize new shard: {:?}", shard);
-		init_shard(&shard)?;
-		Stf::init_state()
-	};
-	trace!("Successfully loaded or initialized state from shard {:?}", shard);
-	Ok(state)
-}
+use itp_settings::files::SHARDS_PATH;
+use itp_types::ShardIdentifier;
+use log::error;
+use std::{
+	format, fs,
+	path::{Path, PathBuf},
+	string::String,
+	vec::Vec,
+};
 
-pub(crate) fn load(shard: &ShardIdentifier) -> Result<StfState> {
-	// load last state
-	let state_path =
-		format!("{}/{}/{}", SHARDS_PATH, shard.encode().to_base58(), ENCRYPTED_STATE_FILE);
-	trace!("loading state from: {}", state_path);
-	let state_vec = read(&state_path)?;
-
-	// state is now decrypted!
-	let state: StfStateType = match state_vec.len() {
-		0 => {
-			debug!("state at {} is empty. will initialize it.", state_path);
-			Stf::init_state().state
-		},
-		n => {
-			debug!("State loaded from {} with size {}B, deserializing...", state_path, n);
-			StfStateType::decode(&mut state_vec.as_slice())?
-		},
-	};
-	trace!("state decoded successfully");
-	// add empty state-diff
-	let state_with_diff = StfState { state, state_diff: Default::default() };
-	trace!("New state created: {:?}", state_with_diff);
-	Ok(state_with_diff)
-}
+/// Trait to abstract file I/O for state.
+pub trait StateFileIo {
+	type StateType;
+	type HashType;
+
+	/// Load a state file (returns error if it does not exist).
```

Comment:
So after our discussion on Friday, I changed it up a bit (because I think you're right on some important points here). The snapshot repository does not know anything about file and folder names now. It knows shard identifiers and state snapshot identifiers. The `FileIo` is then responsible for translating these IDs into folder and file names.